### PR TITLE
Remove CDN config from API

### PR DESCRIPTION
**Pull request by dkarnutsch**
Also: provide helper to ensure traffic is coming from CDN

### CODE REVIEW

File: packages/api/cms-api/src/common/guards/header.guard.ts [hunk 1]
```diff
@@ -0,0 +1,26 @@
+import { CanActivate, ExecutionContext, Injectable } from "@nestjs/common";
+import { GqlExecutionContext } from "@nestjs/graphql";
+import { Request } from "express";
+
+@Injectable()
+/** Blocks all requests without specified header */
+export class CometHeaderGuard implements CanActivate {
```
Comment (nsams) [inline packages/api/cms-api/src/common/guards/header.guard.ts:7]: HeaderGuard is unspecific, what about:

- BlockRequestsWithoutHeader (using your comment)
- BlockWithoutHeader

**Comment by nsams** [comment]
:+1:

should we add this to starter?

File: packages/api/cms-api/src/common/guards/header.guard.ts (hunk 1 above)
Comment (dkarnutsch) [reply packages/api/cms-api/src/common/guards/header.guard.ts]: BlockWithoutHeader sounds good and makes the comment useless :)

**Comment by dkarnutsch** [comment]
> 👍
> 
> should we add this to starter?

yes

File: demo/api/src/main.ts [hunk 2]
```diff
@@ -41,6 +41,11 @@ async function bootstrap(): Promise<void> {
     app.use(compression());
     app.use(cookieParser());
 
+    // if cdn is enabled, make sure all traffic is coming from the cdn
+    if (config.cdn.enabled) {
+        app.useGlobalGuards(new BlockWithoutHeaderGuard({ headerName: "x-cdn-origin-check", headerValue: config.cdn.originCheck }));
+    }
```
Comment (johnnyomair) [inline demo/api/src/main.ts:47]: How would this work for cluster-internal requests that bypass the CDN? For instance, the site querying during build?

Comment (dkarnutsch) [reply demo/api/src/main.ts:47]: Nice catch. We could either attach the header during build (ugly), use a second header (also ugly) or detect that the request is coming from an internal source.

Comment (dkarnutsch) [reply demo/api/src/main.ts:47]: @nsams what do you think?

Comment (dkarnutsch) [reply demo/api/src/main.ts]: I've added a check. Basically, if `x-forwarded-for` is present, we assume the request is coming from outside. This is safe enough, if we trust the reverse proxy.

File: demo/api/src/main.ts [hunk 3]
```diff
@@ -43,6 +43,11 @@ async function bootstrap(): Promise<void> {
     app.use(compression());
     app.use(cookieParser());
 
+    // if CDN is enabled, make sure all traffic is either coming from the CDN or internal sources
+    if (config.cdn.enabled) {
```
Comment (nsams) [inline demo/api/src/main.ts:47]: could you replace that with `config.cdn.originCheck`? And remove cnd.enabled?

File: packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+    private headerValue: string;
+
+    constructor({ headerName, headerValue }: { headerName: string; headerValue: string }) {
+        this.headerName = headerName;
+        this.headerValue = headerValue;
+    }
+
+    private isRequestInternal(request: Request): boolean {
+        // This check is only ok when the application is behind a trusted proxy
+        return !request.header("x-forwarded-for");
```
Comment (nsams) [inline packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts:17]: that works in kubernetes where the api pod is behind a haproxy for external requests, and accessed directly for internal requests (BFF), right?

File: demo/api/src/main.ts (hunk 3 above)
Comment (dkarnutsch) [reply demo/api/src/main.ts:47]: I've done this on purpose to enable easier toggling of this feature.

File: packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts (hunk 4 above)
Comment (dkarnutsch) [reply packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts:17]: Yes exactly. Technically, it is a nginx proxy. But this should work for any trusted proxy.

File: demo/api/src/main.ts (hunk 3 above)
Comment (nsams) [reply demo/api/src/main.ts:47]: imho less config options and env vars should be preferred if possible

Comment (dkarnutsch) [reply demo/api/src/main.ts:47]: I think a more explicit approach is better for easier understanding

@johnnyomair what do you think?

File: packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts (hunk 4 above)
Comment (thomasdax98) [inline packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts:17]: How is this secure? Couldn't anyone add the header to an API request? I suppose the CDN takes care of removing this header from incoming requests? 

Is there any scenario where the API can still be reached directly bypassing the CDN?

File: packages/api/cms-api/src/dam/files/files.service.ts [hunk 5]
```diff
@@ -559,8 +559,7 @@ export class FilesService {
     async createFileUrl(file: FileInterface, previewDamUrls?: boolean): Promise<string> {
         const filename = parse(file.name).name;
 
-        // Use CDN url only if available and not in preview as preview requires auth
-        const baseUrl = [this.config.cdnEnabled && !previewDamUrls ? `${this.config.cdnDomain}/files` : this.config.filesBaseUrl];
+        const baseUrl = [this.config.filesBaseUrl];
```
Comment (thomasdax98) [inline packages/api/cms-api/src/dam/files/files.service.ts:562]: This also removes the previewDamUrls check. Meaning there is no possibility to switch between the privateApiUrl (Authproxy) and publicApiUrl. Will this be readded in a follow-up PR? Or will this be handled differently?

File: demo/api/src/config/config.ts [hunk 6]
```diff
@@ -48,6 +48,10 @@ export function createConfig(processEnv: NodeJS.ProcessEnv) {
             },
             storageDirectoryPrefix: envVars.BLOB_STORAGE_DIRECTORY_PREFIX,
         },
+        cdn: {
+            enabled: envVars.CDN_ENABLED,
+            originCheck: envVars.CDN_ORIGIN_CHECK,
```
[the remark this answers is not in the document]
Comment (dkarnutsch) [reply demo/api/src/config/config.ts:53]: yes, exactly. You configure the CDN to attach this value on every request

File: packages/api/cms-api/src/dam/files/files.service.ts (hunk 5 above)
Comment (dkarnutsch) [reply packages/api/cms-api/src/dam/files/files.service.ts]: Yes, you are absolutely right. I've tried to make as few changes as possible, but it seems that more changes are needed.

I've addressed this in 5e707ebbce4adf879fccc35bda932404608d13fd

Please have another look. Once we agree on a solution, I will update the docs and migration guide accordingly.

File: demo/api/src/main.ts (hunk 3 above)
Comment (dkarnutsch) [reply demo/api/src/main.ts:47]: as discussed, we remove enabled

File: packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts (hunk 4 above)
Comment (dkarnutsch) [reply packages/api/cms-api/src/common/guards/external-request-without-header.guard.ts:17]: as discussed, enable/disable per boolean in code

File: demo/api/src/config/environment-variables.ts [hunk 7]
```diff
@@ -92,4 +95,14 @@ export class EnvironmentVariables {
...
     S3_BUCKET: string;
+
+    @IsOptional()
+    @IsBoolean()
+    @Transform(({ value }) => value === "true")
+    CDN_ENABLED: boolean;
+
+    @IsString()
+    @IsNotEmpty()
+    @ValidateIf((v) => v.CDN_ENABLED)
```
Comment (dkarnutsch) [inline demo/api/src/config/environment-variables.ts:106]: validateIf node_env: production

**Comment by johnnyomair** [comment]
Marking as draft until the changes discussed in last week's meeting have been implemented. Feel free to request a review again then. Thanks!

File: packages/api/cms-api/src/dam/files/files.service.ts [hunk 8]
```diff
@@ -556,11 +556,13 @@ export class FilesService {
...
-    async createFileUrl(file: FileInterface, previewDamUrls?: boolean): Promise<string> {
+    async createFileUrl(
+        file: FileInterface,
+        { previewDamUrls = false, relativeDamUrls = false }: { previewDamUrls?: boolean; relativeDamUrls?: boolean },
+    ): Promise<string> {
         const filename = parse(file.name).name;
 
-        // Use CDN url only if available and not in preview as preview requires auth
-        const baseUrl = [this.config.cdnEnabled && !previewDamUrls ? `${this.config.cdnDomain}/files` : this.config.filesBaseUrl];
+        const baseUrl = [`${relativeDamUrls ? "" : this.config.apiUrl}dam/files`];
```
[the remark this answers is not in the document]
Comment (johnnyomair) [reply packages/api/cms-api/src/dam/files/files.service.ts:565]: Ah, didn't see that.

**Review by johnnyomair — APPROVED** [review]

**Pull request merged**